### PR TITLE
Add error code for misconfigured webhooks

### LINKED ISSUE #5778: Add ERROR_CODE to failed MaintenancePreconditionsSatisfied constraint
**How to categorize this issue?**
/kind enhancement

**What would you like to be added**:
In order to make the user aware of misconfigured webhooks we want to show this as user action error in the dashboard: https://github.com/gardener/dashboard/issues/1144
It would be great, if the reason for the failed constraint could be added as error code, like this:

```yaml
constraints:
  - type: MaintenancePreconditionsSatisfied
    status: 'False'
    codes:
      - ERR_USER_WEBHOOK
    lastTransitionTime: '2022-04-08T03:59:02Z'
    lastUpdateTime: '2022-04-07T15:59:58Z'
    reason: ProblematicWebhooks
    message: >-
      MutatingWebhookConfiguration "fail-webhook" is problematic: webhook
      "my-webhook.example.com" with failurePolicy "Fail" and 5s timeout might
      prevent worker nodes from properly joining the shoot cluster
```

Error codes are already added to conditions in some cases. Adding an error code also to constraints would help us to make the code universal and ready to support future additions.

**Why is this needed**:
The Dashboard is supposed to show this as user error to make the user aware that he needs to take action. However, we can only guess that a webhook is the reason. Gardener should introduce `ERR_USER_WEBHOOK` (or a similar error code) and add it to the failed constraint. Maybe there can be other reasons for this condition to fail which we can introduce at a later point.

### PR DESCRIPTION
**Pull request by ary1992**
**How to categorize this PR?**

/kind enhancement

**What this PR does / why we need it**:
In case of misconfigured webhooks , the shoot status constraints fails with reason `ProblematicWebhooks`. This PR adds the error code to reason for the failed constraint.

**Which issue(s) this PR fixes**:
Fixes https://github.com/gardener/gardener/issues/5778

**Special notes for your reviewer**:

**Release note**:

```other operator

```

### CODE REVIEW

File: pkg/apis/core/types_common.go [hunk 1]
```diff
@@ -47,6 +47,8 @@ const (
 	ErrorConfigurationProblem ErrorCode = "ERR_CONFIGURATION_PROBLEM"
 	// ErrorRetryableConfigurationProblem indicates that the last error occurred due to a retryable configuration problem.
 	ErrorRetryableConfigurationProblem ErrorCode = "ERR_RETRYABLE_CONFIGURATION_PROBLEM"
+	// ErorUserWebhook indicates that the last error occurred due to misconfigurred webhook.
```
Comment (shafeeqes) [inline pkg/apis/core/types_common.go]:
```suggestion
	// ErorUserWebhook indicates that the last error occurred due to misconfigured webhook.
```
Please change at other places too.

File: pkg/operation/care/constraints.go [hunk 2]
```diff
@@ -171,10 +171,11 @@ func (c *Constraint) CheckForProblematicWebhooks(ctx context.Context) (gardencor
 	for _, webhookConfig := range validatingWebhookConfigs {
 		for _, w := range webhookConfig.Webhooks {
 			if IsProblematicWebhook(w.FailurePolicy, w.ObjectSelector, w.NamespaceSelector, w.Rules, w.TimeoutSeconds) {
+				msg := buildProblematicWebhookMessage("ValidatingWebhookConfiguration", webhookConfig.Name, w.Name, w.FailurePolicy, w.TimeoutSeconds)
 				return gardencorev1beta1.ConditionFalse,
 					"ProblematicWebhooks",
-					buildProblematicWebhookMessage("ValidatingWebhookConfiguration", webhookConfig.Name, w.Name, w.FailurePolicy, w.TimeoutSeconds),
-					nil
+					msg,
+					gardencorev1beta1helper.NewErrorWithCodes(msg, gardencorev1beta1.ErrorUserWebhook)
```
Comment (shafeeqes) [inline pkg/operation/care/constraints.go]: I think you'll also need to make changes here,
https://github.com/gardener/gardener/blob/f7b388e2db1256de4b2254246e7cff7f177946e3/pkg/operation/care/checker.go#L648
`return gardencorev1beta1helper.UpdatedConditionUnknownError(oldCondition, err, gardencorev1beta1helper.ExtractErrorCodes(err)...)`

**Review by shafeeqes — CHANGES_REQUESTED** [review]

File: pkg/operation/care/constraints.go (hunk 2 above)
Comment (ary1992) [reply pkg/operation/care/constraints.go]: Thank you for the feedback. I have made the changes as per suggestion. I will test the changes as well.

File: docs/usage/shoot_status.md [hunk 3]
```diff
@@ -105,3 +105,4 @@ Known error codes are:
 - `ERR_CLEANUP_CLUSTER_RESOURCES` - indicates that the last error occurred due to resources in the cluster that are stuck in deletion.
 - `ERR_CONFIGURATION_PROBLEM` - indicates that the last error occurred due to a configuration problem. It is classified as a non-retryable error code.
 - `ERR_RETRYABLE_CONFIGURATION_PROBLEM` - indicates that the last error occurred due to a retryable configuration problem. "Retryable" means that the occurred error is likely to be resolved in a ungraceful manner after given period of time.
+- `ERR_USER_WEBHOOK` - indicates that the last error occured due to misconfigurred webhook.
```
Comment (acumino) [inline docs/usage/shoot_status.md]:
```suggestion
- `ERR_USER_WEBHOOK` - indicates that the last error occurred due to misconfigured webhook.
```

File: pkg/operation/care/constraints.go (hunk 2 above)
Comment (ary1992) [reply pkg/operation/care/constraints.go]: In the [commit](https://github.com/gardener/gardener/pull/5855/commits/ef5e8575d0a9a41eedf5a0e8a3c1a0732901a8c4), we are returning error codes from `CheckForProblematicWebhooks` function. This what I have got after testing the implementation in local setup.

Case : Problematic webhook present
```
Constraints:
    Codes:
      ERR_USER_WEBHOOK
    Last Transition Time:  2022-04-28T10:15:50Z
    Last Update Time:      2022-04-28T10:15:43Z
    Message:               MutatingWebhookConfiguration "fail-webhook" is problematic: webhook "my-webhook.example.com" with failurePolicy "Fail" and 5s timeout might prevent worker nodes from properly joining the shoot cluster
    Reason:                ProblematicWebhooks
    Status:                False
    Type:                  HibernationPossible
    Codes:
      ERR_USER_WEBHOOK
    Last Transition Time:  2022-04-28T10:15:50Z
    Last Update Time:      2022-04-28T10:15:43Z
    Message:               MutatingWebhookConfiguration "fail-webhook" is problematic: webhook "my-webhook.example.com" with failurePolicy "Fail" and 5s timeout might prevent worker nodes from properly joining the shoot cluster
    Reason:                ProblematicWebhooks
    Status:                False
    Type:                  MaintenancePreconditionsSatisfied
```

Case: Problematic webhook not present
```
Constraints:
    Last Transition Time:  2022-04-28T10:46:56Z
    Last Update Time:      2022-04-28T10:46:14Z
    Message:               All webhooks are properly configured.
    Reason:                NoProblematicWebhooks
    Status:                True
    Type:                  HibernationPossible
    Last Transition Time:  2022-04-28T10:46:56Z
    Last Update Time:      2022-04-28T10:46:14Z
    Message:               All webhooks are properly configured.
    Reason:                NoProblematicWebhooks
    Status:                True
    Type:                  MaintenancePreconditionsSatisfied
```

Is this the right implementation ? @timebertt @grolu

File: pkg/apis/core/types_common.go [hunk 4]
```diff
@@ -47,6 +47,8 @@ const (
 	ErrorConfigurationProblem ErrorCode = "ERR_CONFIGURATION_PROBLEM"
 	// ErrorRetryableConfigurationProblem indicates that the last error occurred due to a retryable configuration problem.
 	ErrorRetryableConfigurationProblem ErrorCode = "ERR_RETRYABLE_CONFIGURATION_PROBLEM"
+	// ErrorUserWebhook indicates that the last error occurred due to misconfigured webhook.
+	ErrorUserWebhook ErrorCode = "ERR_USER_WEBHOOK"
 )
```
Comment (rfranzke) [inline pkg/apis/core/types_common.go]:
```suggestion
	ErrorProblematicWebhook ErrorCode = "ERR_PROBLEMATIC_WEBHOOK"
)
```

File: pkg/apis/core/types_common.go [hunk 5]
```diff
@@ -47,6 +47,8 @@ const (
 	ErrorConfigurationProblem ErrorCode = "ERR_CONFIGURATION_PROBLEM"
 	// ErrorRetryableConfigurationProblem indicates that the last error occurred due to a retryable configuration problem.
 	ErrorRetryableConfigurationProblem ErrorCode = "ERR_RETRYABLE_CONFIGURATION_PROBLEM"
+	// ErrorUserWebhook indicates that the last error occurred due to misconfigured webhook.
```
Comment (rfranzke) [inline pkg/apis/core/types_common.go]:
```suggestion
	// ErrorUserWebhook indicates that the last error occurred due to a webhook not following the Kubernetes best practices (https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#best-practices-and-warnings).
```

File: docs/usage/shoot_status.md [hunk 6]
```diff
@@ -105,3 +105,4 @@ Known error codes are:
 - `ERR_CLEANUP_CLUSTER_RESOURCES` - indicates that the last error occurred due to resources in the cluster that are stuck in deletion.
 - `ERR_CONFIGURATION_PROBLEM` - indicates that the last error occurred due to a configuration problem. It is classified as a non-retryable error code.
 - `ERR_RETRYABLE_CONFIGURATION_PROBLEM` - indicates that the last error occurred due to a retryable configuration problem. "Retryable" means that the occurred error is likely to be resolved in a ungraceful manner after given period of time.
+- `ERR_USER_WEBHOOK` - indicates that the last error occurred due to misconfigured webhook.
```
Comment (rfranzke) [inline docs/usage/shoot_status.md]:
```suggestion
- `ERR_PROBLEMATIC_WEBHOOK` - indicates that the last error occurred due to a webhook not following the Kubernetes best practices (https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#best-practices-and-warnings).
```

File: pkg/apis/core/v1alpha1/types_common.go [hunk 7]
```diff
@@ -45,6 +45,9 @@ const (
 	ErrorConfigurationProblem ErrorCode = "ERR_CONFIGURATION_PROBLEM"
 	// ErrorRetryableConfigurationProblem indicates that the last error occurred due to a retryable configuration problem.
 	ErrorRetryableConfigurationProblem ErrorCode = "ERR_RETRYABLE_CONFIGURATION_PROBLEM"
+	// ErrorUserWebhook indicates that the last error occurred due to misconfigured webhook.
+	// It is classified as a non-retryable error code.
+	ErrorUserWebhook ErrorCode = "ERR_USER_WEBHOOK"
```
Comment (rfranzke) [inline pkg/apis/core/v1alpha1/types_common.go]: see above

File: pkg/apis/core/v1beta1/types_common.go [hunk 8]
```diff
@@ -45,6 +45,8 @@ const (
 	ErrorConfigurationProblem ErrorCode = "ERR_CONFIGURATION_PROBLEM"
 	// ErrorRetryableConfigurationProblem indicates that the last error occurred due to a retryable configuration problem.
 	ErrorRetryableConfigurationProblem ErrorCode = "ERR_RETRYABLE_CONFIGURATION_PROBLEM"
+	// ErrorUserWebhook indicates that the last error occurred due to misconfigured webhook.
+	ErrorUserWebhook ErrorCode = "ERR_USER_WEBHOOK"
```
Comment (rfranzke) [inline pkg/apis/core/v1beta1/types_common.go]: see above

File: pkg/apis/core/v1alpha1/types_common.go (hunk 7 above)
Comment (ary1992) [reply pkg/apis/core/v1alpha1/types_common.go]: Done in [commit](https://github.com/gardener/gardener/pull/5855/commits/af3682c50b74375e16ce22713e6477f41c56920c).

File: pkg/apis/core/v1beta1/types_common.go (hunk 8 above)
Comment (ary1992) [reply pkg/apis/core/v1beta1/types_common.go]: Done in [commit](https://github.com/gardener/gardener/pull/5855/commits/af3682c50b74375e16ce22713e6477f41c56920c).

**Review by rfranzke — COMMENTED** [review]
Nice!
/approve

File: pkg/operation/care/constraints.go (hunk 2 above)
Comment (shafeeqes) [reply pkg/operation/care/constraints.go]: Nit: you can mention this in PR description also.

**Pull request merged**